eth/db/storage.py

```diff
@@ -0,0 +1,477 @@
+    Similarly, it persists changes to the appropriate trie at write time.
+
+    StorageLookup also tracks the state roots changed since the last persist.
+    """
+    logger = cast(ExtendedDebugLogger, logging.getLogger("eth.db.storage.StorageLookup"))
+
+    def __init__(self, db: BaseDB, account_db: BaseAccountDB, address: Address) -> None:
+        self._db = db
+        self._account_db = account_db
+        self._address = address
+        self._write_trie = None
+        self._starting_root_hash = None  # type: Hash32
+        self._trie_nodes_batch = None  # type: BatchDB
+
+    def get_storage_root_at_last_persist(self) -> Hash32:
+        if self._starting_root_hash is None:
+            raise ValidationError(
+                "Tried to get original storage root for %s before it was available" % (
+                    self._address,
+                )
+            )
+        else:
+            return self._starting_root_hash
+
+    def _get_write_trie(self) -> HexaryTrie:
+        if self._trie_nodes_batch is None:
+            self._trie_nodes_batch = BatchDB(self._db, read_through_deletes=True)
+
+        if self._write_trie is None:
+            storage_root = self._account_db.get_storage_root(self._address)
+            batched_db = self._trie_nodes_batch
+            self._write_trie = HexaryTrie(batched_db, root_hash=storage_root, prune=True)
+            if self._starting_root_hash is None:
+                self._starting_root_hash = storage_root
+
+        return self._write_trie
+
+    def _get_read_trie(self) -> HexaryTrie:
+        storage_root = self._account_db.get_storage_root(self._address)
+        if storage_root == BLANK_ROOT_HASH:
+            return HexaryTrie({})
+        elif self._write_trie is not None:
+            return self._write_trie
+        else:
+            if self._starting_root_hash is None:
+                self._starting_root_hash = storage_root
+            # Creating "HexaryTrie" is a pretty light operation, so not a huge cost
+            # to create a new one at every read, but we could
+            # cache the read trie, if this becomes a bottleneck.
+            return HexaryTrie(self._db, root_hash=storage_root)
+
+    def _decode_key(self, key: bytes) -> bytes:
+        padded_slot = pad32(key)
+        return keccak(padded_slot)
+
+    def __getitem__(self, key: bytes) -> bytes:
+        hashed_slot = self._decode_key(key)
+        read_trie = self._get_read_trie()
+        try:
+            return read_trie[hashed_slot]
+        except trie_exceptions.MissingTrieNode as exc:
+            self.logger.warning(exc)
+            self.logger.debug("expected to get slot, but:", exc_info=True)
+            raise exc
+
+    def __setitem__(self, key: bytes, value: bytes) -> None:
+        hashed_slot = self._decode_key(key)
+        write_trie = self._get_write_trie()
+        write_trie[hashed_slot] = value
+
+    def _exists(self, key: bytes) -> bool:
```

Comment:
This method is also unused.

Reply:
This one is a little tricky, it's actually used by `BaseDB` for `__contains__` checks. I can add an inline comment.

Reply:
I suppose we're not doing any explicit `in` checks right now, but I'm comfortable just having this implemented so that it will work when someone needs it.

Reply:
Ahh, good point, I checked for `in` but missed that `BaseDB` actually uses this. `BaseDB` doesn't just call `__contains__` because then everyone would need to add `# type: ignore`?

Reply:
Thanks for the comment! That will help future Brian's :)